### ISSUE TITLE
Where to use `to(self.device)` ?

### ISSUE THREAD

**Issue by bicycleman15** [PyTorchLightning/pytorch-lightning#2638]
#### What is your question?
So currently in my lightning module, I create the multibox module inside the `training_step`, in order to send the priorboxes to the right device, but when I define the multibox loss in the `__init__` method, the priors doesn't seem to be in the right device.

#### Code
for eg. below **works**:-
```python
class SSD_simple(pl.LightningModule):

    def __init__(self,config : dict):
        super().__init__()
        self.config = config

        self.model = SSD300()

    
    def forward(self, x):
        return self.model(x)
    
    def training_step(self, batch, batch_nb):
        images, bboxes, labels = batch
        locs, confs = self(images)

        priors = PriorBox(self.config)
        priors = priors.create_priors().to(self.device)
        multibox_loss = MultiBoxLoss(num_classes=self.config['num_classes'],priors=priors,cfg=self.config,device=self.device)

        loc_loss, conf_loss = multibox_loss(locs, confs, bboxes, labels)
        loss = conf_loss + loc_loss

        logs = {'train_loss' : loss, 'conf_loss':conf_loss, 'loc_loss' : loc_loss}

        return {'loss' : loss, 'log' : logs}
```
But the below **doesn't work** :-
```python
class SSD_simple(pl.LightningModule):

    def __init__(self,config : dict):
        super().__init__()
        self.config = config

        self.model = SSD300()

        priors = PriorBox(self.config)
        priors = priors.create_priors().to(self.device)
        self.multibox_loss = MultiBoxLoss(num_classes=self.config['num_classes'],priors=priors,cfg=self.config,device=self.device)

    
    def forward(self, x):
        return self.model(x)
    
    def training_step(self, batch, batch_nb):
        images, bboxes, labels = batch
        locs, confs = self(images)

        loc_loss, conf_loss = self.multibox_loss(locs, confs, bboxes, labels)
        loss = conf_loss + loc_loss

        logs = {'train_loss' : loss, 'conf_loss':conf_loss, 'loc_loss' : loc_loss}

        return {'loss' : loss, 'log' : logs}
```
It throws the following error:
`RuntimeError: expected device cuda:0 but got device cpu`

I want to get the later method to work since it is not computationally expensive as I am not creating priors again and again.

Any help ? I was thinking whether there was a pre_training method ?

#### What's your environment?
 - OS: [e.g. Linux]
 - Packaging [pip]

**Comment by SurajDonthi** [comment]
Hey @bicycleman15,

Were you able to resolve this issue? Please share the details.

**Comment by bicycleman15** [comment]
@SurajDonthi  yeah so the thing is when `__init__` is called, the model is not on the right device initially, but it is sent later on. That is why when `__init__` is called, `self.device` is `cpu`, but later on when you will check, it will eventually end up on the right device during training.

I fixed this by registering priors in buffers of the `SSD300` model class so that they automatically sent to the right device along the model (by using `self.register_buffer`).

Hope this helps.

**Comment by SurajDonthi** [comment]
@bicycleman15,

Agreed. Inferred the same and made changes to the code accordingly!

**Comment by wisecornelius** [comment]
What is the first `LightningHook` where the module is on the correct device? I thought that would be `setup` but it is still on the `cpu` there.

**Comment by jchia** [comment]
@wisecornelius In `setup()`, `self.device` should be correct after this issue is fixed: #13108 13108

**Comment by yannqi** [comment]
The suggest answer is here: https://lightning.ai/docs/pytorch/stable/accelerators/accelerator_prepare.html
